File: test/testfwk/xdevice/src/xdevice/_core/context/handler.py
#!/usr/bin/env python3
# coding=utf-8

#
# Copyright (c) 2020-2023 Huawei Device Co., Ltd.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import os
import shutil
import time
import json
from xml.etree import ElementTree

from _core.constants import CaseResult
from _core.constants import FilePermission
from _core.context.upload import Uploader
from _core.logger import platform_logger
from _core.report.reporter_helper import ReportConstant

__all__ = [
    "handle_repeat_result", "update_report_xml", "report_not_executed"
]

LOG = platform_logger("Context")


def handle_repeat_result(report_xml, report_path, round_folder=""):
    if not round_folder:
        return report_xml
    round_path = os.path.join(report_path, "result", round_folder)
    if not os.path.exists(round_path):
        os.makedirs(round_path)
    LOG.debug("move result file to round folder")
    target_path = os.path.join(round_path, os.path.basename(report_xml))
    shutil.move(report_xml, target_path)
    return target_path


def update_report_xml(report_xml, props):
    """update devices, start time, end time, etc. to the result file"""
    if not os.path.exists(report_xml) or not props:
        return
    try:
        root = ElementTree.parse(report_xml).getroot()
    except ElementTree.ParseError as e:
        LOG.error(f"parse result xml error! xml file {report_xml}")
        LOG.error(f"error message: {e}")
        return
    for k, v in props.items():
        if k == ReportConstant.devices:
            v = f"<![CDATA[{json.dumps(v, separators=(',', ':'))}]]>"
        root.set(k, v)
    result_fd = os.open(report_xml, os.O_CREAT | os.O_WRONLY | os.O_TRUNC, FilePermission.mode_644)
    with os.fdopen(result_fd, mode="w", encoding="utf-8") as result_file:
        result_file.write(ElementTree.tostring(root).decode())


def report_not_executed(report_path, test_drivers, error_message, task=None):
    from _core.utils import check_result_report
    from _core.utils import get_repeat_round
    from _core.utils import get_sub_path

    repeat, repeat_round = 1, 1
    if task is not None:
        repeat = task.config.repeat

    for test_driver in test_drivers:
        _, test = test_driver
        module_name = test.source.module_name
        test_name = test.source.test_name
        repeat_round = get_repeat_round(test.unique_id)
        round_folder = f"round{repeat_round}" if repeat > 1 else ""

        # get report file
        if task and getattr(task.config, "testdict", ""):
            report_file = os.path.join(get_sub_path(test.source.source_file), "%s.xml" % test_name)
        else:
            report_file = os.path.join(report_path, "result", round_folder, "%s.xml" % module_name)

        # get report name
        report_name = test_name if not test_name.startswith("{") else "report"
        # here, normally create empty report and then upload result
        check_result_report(report_path, report_file, error_message, report_name, module_name,
                            result_kind=CaseResult.unavailable)

        update_props = {
            ReportConstant.start_time: time.strftime(ReportConstant.time_format, time.localtime()),
            ReportConstant.end_time: time.strftime(ReportConstant.time_format, time.localtime()),
            ReportConstant.repeat: str(repeat),
            ReportConstant.round: str(repeat_round),
            ReportConstant.test_type: test.source.test_type
        }
        update_report_xml(report_file, update_props)
        Uploader.upload_unavailable_result(
            error_message,
            case_id=module_name,
            report_path=os.path.join(report_path, "log", ReportConstant.task_run_log)
        )

File: foundation/CastEngine/castengine_cast_plus_stream/src/utils/src/handler.cpp
/*
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * Description: handler module for send message and handle message in a looper
 * Author: lijianzhao
 * Create: 2022-01-19
 */

#include "handler.h"

namespace OHOS {
namespace CastEngine {
namespace CastEngineService {
Handler::Handler() : stop_(false), stopWhenEmpty_(false)
{
    looper_ = std::thread([this]() {
        for (;;) {
            Message msg;
            {
                std::unique_lock<std::mutex> lock(this->queueMutex_);
                if ((this->stopWhenEmpty_ && this->msgQueue_.empty()) || this->stop_) {
                    return;
                }

                if (this->msgQueue_.empty()) {
                    this->condition_.wait(lock);
                }

                if ((this->stopWhenEmpty_ && this->msgQueue_.empty()) || this->stop_) {
                    return;
                }

                // send message at when_
                if (!(this->msgQueue_.empty()) &&
                    (this->condition_.wait_until(lock, this->msgQueue_.back().when_) != std::cv_status::timeout)) {
                    continue;
                }

                if (!this->msgQueue_.empty()) {
                    msg = std::move(this->msgQueue_.back());
                    this->msgQueue_.pop_back();
                } else {
                    continue;
                }
            }
            this->HandleMessageInner(msg);
        }
    });
}

Handler::~Handler()
{
    {
        std::unique_lock<std::mutex> lock(queueMutex_);
        stop_ = true;
        condition_.notify_all();
    }

    if (looper_.joinable()) {
        looper_.join();
    }

    msgQueue_.clear();
}

void Handler::ThreadJoin()
{
    if (looper_.joinable()) {
        looper_.join();
    }
}

bool Handler::SendCastMessage(const Message &msg)
{
    std::unique_lock<std::mutex> lock(queueMutex_);
    auto i = std::find(msgQueue_.begin(), msgQueue_.end(), msg);
    if (i != msgQueue_.end()) {
        msgQueue_.erase(i);
    }

    msgQueue_.push_back(msg);
    std::sort(msgQueue_.begin(), msgQueue_.end(), std::greater<Message>());
    condition_.notify_one();
    return true;
}

bool Handler::SendCastMessage(int what)
{
    return SendCastMessage(Message(what));
}

bool Handler::SendCastMessage(int what, int arg1)
{
    return SendCastMessage(Message(what, arg1));
}

bool Handler::SendCastMessage(int what, int arg1, int arg2)
{
    return SendCastMessage(Message(what, arg1, arg2));
}

bool Handler::SendCastMessageDelayed(int what, long uptimeMillis)
{
    if (uptimeMillis < 0) {
        return false;
    }

    Message msg(what);
    msg.SetWhen(uptimeMillis);
    return SendCastMessage(msg);
}

void Handler::RemoveMessage(const Message &msg)
{
    std::unique_lock<std::mutex> lock(queueMutex_);
    for (auto it = msgQueue_.begin(); it != msgQueue_.end();) {
        if (it->what_ == msg.what_) {
            it = msgQueue_.erase(it);
        } else {
            ++it;
        }
    }
    condition_.notify_one();
}

void Handler::RemoveCallbackAndMessages()
{
    std::unique_lock<std::mutex> lock(queueMutex_);
    msgQueue_.clear();
}

void Handler::StopSafty(bool stopSafty)
{
    std::unique_lock<std::mutex> lock(queueMutex_);
    if (stopSafty) {
        stopWhenEmpty_ = true;
    } else {
        stop_ = true;
    }

    condition_.notify_all();
}

bool Handler::IsQuiting()
{
    std::unique_lock<std::mutex> lock(queueMutex_);
    if (stop_ || stopWhenEmpty_) {
        return true;
    }
    return false;
}

void Handler::HandleMessageInner(const Message &msg)
{
    if (msg.task_ != nullptr) {
        msg.task_();
    } else {
        if (msg.what_ < 0) {
            return;
        }
        HandleMessage(msg);
    }
}
} // namespace CastEngineService
} // namespace CastEngine
} // namespace OHOS
